### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by tgross**
Includes working Dockerfile and Vault configuration, but the unsealing and ACL creation process still needs a bunch of work. Another TODO will be figuring how to get TLS certs deployed for this.

cc @misterbisson this is still work-in-progress.

### CODE REVIEW

File: setup.sh [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    docker exec -it vault_consul-vault_1 vault init \
+           -address='http://127.0.0.1:8200' \
+           -key-shares=${#keys[@]} \
+           -key-threshold=2 \
+           -pgp-keys="${1}" > vault.keys 2>&1 \
+    && echo 'Vault initialized. Distribute ./vault.keys as required.'
+}
+
+# use the first two encrypted keys downloaded in vault.keys to unseal all vault nodes
+# TODO: this really should be done by multiple people
```
Comment (tgross) [inline setup.sh]: @misterbisson I'd love to have your thoughts on this.

Right now the process is to get PGP public keys from your operators, `docker-compose up -d`, wait for the cluster to be healthy, and then run `./setup.sh init 'key1.asc,key2.asc,key3.asc'`. This creates a `vault.keys` file in which you can only decrypt one of the keys (your own). You'll need to distribute chunks of this file out-of-band to your operators. The unsealing process requires a minimum of 2 keys if you have more than one key in the split secret (i.e. if you have 3 shards, you can get away with 2 keys to unseal).

There's a tension here between having this blueprint be easy to demonstrate and having it be something someone could translate into a production process. My suggestion is to keep `./setup.sh init` as is, provide `./setup.sh unseal` that operates only on a single key, and then have a `./setup.sh demo` that creates and inits with two keys and then unseals with both of them on a single user's laptop. This way we can show how you'd do it in theory but still have a demonstration. What do you think?

Comment (misterbisson) [reply setup.sh]: > This way we can show how you'd do it in theory but still have a demonstration. What do you think?

I think this makes sense as a working plan.

**Comment by tgross** [comment]
@misterbisson something I keep running into while I'm working on this is related to https://github.com/autopilotpattern/consul/issues/25. Vault and Consul both assume that you've got the TLS cert handy at startup. Consul can't get the cert from Vault because Vault requires Consul to be up already if it's going to serve as the backend. Fortunately both accept reload on SIGHUP, so I'm going to stand up the cluster and then distribute the certs out-of-band via `docker exec`, similarly how we're adding the operator keys.

Of course this means our demo also needs a simple certificate authority.

**Comment by tgross** [comment]
Only bits left to do here is to get the gossip key and TLS cert installed after startup. Probably want to do both before we unseal the Vault (so that we don't have to unseal it again after the SIGHUP), but then add both to the Vault afterwards so that application `preStart` hooks can snag them from Vault and use them for setting up their own Consul agents.

**Comment by tgross** [comment]
Apparently we can't use SIGHUP to gracefully reload. We actually have to restart the whole container, alas. See [Reloadable Configuration (bottom of page)](https://www.consul.io/docs/agent/options.html).

**Comment by tgross** [comment]
Got the securing of Consul communication in place. Needs some testing now but is in pretty good shape.

File: setup.sh [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+_demo_wait_for_consul() {
+    echo
+    bold '* Waiting for Consul to form raft...'
+    while :
+    do
+        docker exec -it ${vault}_1 consul info | grep -q "num_peers = 2" && break
+        echo -n '.'
+        sleep 1
+    done
+    reset # something here is breaking the terminal
```
Comment (leoj3n) [inline setup.sh:373]: I've found you can avoid this problem by running docker commands that don't need user input non-interactively, by simply dropping the `-it` as I have done in my version of this script:

https://github.com/leoj3n/consul-vault/blob/08fe1e0492fe871d2885f5509d93f1a2dc80f7e8/setup.sh#L415

**Pull request merged**